src/PureScript/CST/ModuleGraph.purs

```diff
@@ -0,0 +1,117 @@
+import Data.Newtype (un)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Tuple (Tuple(..))
+import Control.Monad.Free (Free, runFree)
+import PureScript.CST.Types (ImportDecl(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (ModuleHeader e) -> Graph ModuleName
+moduleGraph = Map.fromFoldable <<< map go
+  where
+  go (ModuleHeader { name: Name { name }, imports }) =
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { "module": Name { name } }) = name
+
+data ModuleSort e
+  = Sorted (Array (ModuleHeader e))
+  | CycleDetected (Array (ModuleHeader e))
+
+sortModules :: forall e. Array (ModuleHeader e) -> ModuleSort e
+sortModules moduleHeaders = do
+  let
+    knownModuleHeaders :: Map ModuleName (ModuleHeader e)
+    knownModuleHeaders =
+      moduleHeaders
+        # map (\header@(ModuleHeader { name: Name { name } }) -> Tuple name header)
+        # Map.fromFoldable
+
+    graph = moduleGraph moduleHeaders
+    lookupModuleHeaders = Array.mapMaybe (flip Map.lookup knownModuleHeaders) <<< List.toUnfoldable
+
+  case bimap lookupModuleHeaders lookupModuleHeaders (topoSort graph) of
+    Left cycle -> CycleDetected cycle
+    Right sorted -> Sorted sorted
+
+type TopoSortArgs a =
+  { roots :: Set a
+  , sorted :: List a
+  , usages :: Map a Int
+  }
+
+topoSort :: forall a. Ord a => Graph a -> Either (List a) (List a)
+topoSort graph = do
+  _.sorted <$> go { roots: startingModules, sorted: Nil, usages: importCounts }
+  where
+  go :: TopoSortArgs a -> Either (List a) (TopoSortArgs a)
+  go { roots, sorted, usages } = case Set.findMin roots of
+    Nothing ->
+      if all (eq 0) usages then
+        Right { roots, sorted, usages }
+      else do
+        let
+          nonLeaf = Set.fromFoldable $ Map.keys $ Map.filterWithKey (\a count -> count /= 0 && Map.lookup a graph /= Nothing && Map.lookup a graph /= Just Set.empty) usages
+
+        case foldl (\b a -> if isJust b then b else runFree (un Identity) (un Compose (depthFirst { path: Nil, visited: Set.empty, curr: a }))) Nothing nonLeaf of
+          Just cycle -> Left cycle
+          Nothing -> Left Nil
+
+    Just curr -> do
+      let
+        reachable = fromMaybe Set.empty (Map.lookup curr graph)
+        usages' = foldl decrementImport usages reachable
+      go
+        { roots: foldl (appendRoots usages') (Set.delete curr roots) reachable
+        , sorted: Cons curr sorted
+        , usages: usages'
+        }
+
+  appendRoots :: Map a Int -> Set a -> a -> Set a
+  appendRoots usages roots curr = maybe roots (flip Set.insert roots) do
+    count <- Map.lookup curr usages
+    isRoot (Tuple curr count)
+
+  decrementImport :: Map a Int -> a -> Map a Int
+  decrementImport usages k = Map.insertWith add k (-1) usages
+
+  startingModules :: Set a
+  startingModules = Set.fromFoldable $ Map.keys $ Map.filterWithKey (\k v -> isJust (isRoot (Tuple k v))) importCounts
```

Comment:
Unnecessary Set.fromFoldable

Reply:
Huh, I could have sworn `Map.keys` said it returned a `List` on pursuit :confused: